### PR TITLE
Added Total Due Amt/Till Money Amt + Partial Payment Update

### PR DESCRIPTION
**Pull request by WebShells**
Added Total Due Amt to Transaction Summary Reports Depending on a date Selected.
Added Total Due Amt to Customer Specific Report Depending on a date Selected.
Added Due amt to Transactions Report Depending on a date selected/Subtracted from total amount to result in Till's Money Amount.
Support now updating client payment through takings window or transactions report by adding the paid amount and refreshing the payment button.
Will update remaining reports depending on total Due Amnt needs ex: employees report...  
Issues #24 #1811 #1869 #1428

### CODE REVIEW

**Comment by mckaygerhard** [comment]
great! hi @WebShells do you read the last comment about the accounting references in the issue #24 ?

**Comment by WebShells** [comment]
@mckaygerhard Hello Back... Can you be more specific ?

**Comment by mckaygerhard** [comment]
well i cannot explain so well due the languaje (i not speak english i hate it), but the employee/office must have the relation of accounting of what was in credit and was not, the "due" its only "before" the records of the client yet paied are in .. after some records are in, are only "credits" yet paid!, see my comment, due the (very well done) simple behaviour of the payments table, we cannot do that without a extra tables for..

**Comment by mckaygerhard** [comment]
ah sorry @WebShells my bad, it's here what i means: https://github.com/opensourcepos/opensourcepos/issues/1427#issuecomment-371530881 i think that code will be usefully, and remenber a credit amount must not to be changed, this will affects accounting respect the linked sale

File: application/views/sales/form.php [hunk 1]
```diff
@@ -43,19 +43,75 @@
...
 						<?php endif; ?>
 				</div>
 				<div class='col-xs-4'>
 					<div class="input-group input-group-sm">
 						<?php if(!currency_side()): ?>
 							<span class="input-group-addon input-sm"><b><?php echo $this->config->item('currency_symbol'); ?></b></span>
 						<?php endif; ?>
+						<?php if($row->payment_type == "Due") { ?>
+						<?php
+$sql = $this->db->query("SELECT * FROM ospos_sales_payments WHERE sale_id = $row->sale_id LIMIT 1");
```
Comment (jekkos) [inline application/views/sales/form.php]: Some indentation issues here..

File: application/views/sales/form.php [hunk 2]
```diff
@@ -43,19 +43,75 @@
...
+						{	
+				echo "<script>function change(){ var cash = document.getElementById('payment_amount_0'); var debt = document.getElementById('payment_amount_1'); var minus = document.getElementById('VallMinus'); var plus = document.getElementById('VallPlus'); var variable = document.getElementById('minusval'); minus.innerHTML = (debt.value - variable.value); plus.innerHTML = (+cash.value + +variable.value);}
+			 function SubmitChange(){ var cash = document.getElementById('payment_amount_0'); var debt = document.getElementById('payment_amount_1'); var minus = document.getElementById('VallMinus'); var plus = document.getElementById('VallPlus'); var variable = document.getElementById('minusval'); debt.value = (debt.value - variable.value); cash.value = plus.innerHTML;}</script>";
+			 
+						if($row1['payment_type'] == 'Cash' || $row1['payment_type'] == 'Debit Card' || $row1['payment_type'] == 'Credit Card' || $row1['payment_type'] == 'Gift Card' || $row1['payment_type'] == 'Check'){
+						echo form_dropdown('payment_type_'.$i, $payment_options, $row->payment_type, array('id'=>'payment_types_'.$i, 'class'=>'form-control')); 
+						} else {
+			  $j = $i + 1;
```
Comment (jekkos) [inline application/views/sales/form.php]: please fix indentation

File: application/views/sales/form.php [hunk 3]
```diff
@@ -43,19 +43,75 @@
...
-						<?php if( !empty(strstr($row->payment_type, $this->lang->line('sales_giftcard'))) ): ?>
-							<?php echo form_input(array('name'=>'payment_type_'.$i, 'value'=>$row->payment_type, 'id'=>'payment_type_'.$i, 'class'=>'form-control input-sm', 'readonly'=>'true'));?>
-						<?php else: ?>
-							<?php echo form_dropdown('payment_type_'.$i, $payment_options, $row->payment_type, array('id'=>'payment_types_'.$i, 'class'=>'form-control')); ?>
+					<?php // no editing of Gift Card payments as it's a complex change ?>
+					<?php if( !empty(strstr($row->payment_type, $this->lang->line('sales_giftcard'))) ): ?>
+						<?php echo form_input(array('name'=>'payment_type_'.$i, 'value'=>$row->payment_type, 'id'=>'payment_type_'.$i, 'class'=>'form-control input-sm', 'readonly'=>'true'));?>
+					<?php else: ?>
+					<?php 
+						$sqlp = $this->db->query("SELECT * FROM ospos_sales_payments WHERE sale_id = $row->sale_id LIMIT 1");
```
Comment (jekkos) [inline application/views/sales/form.php]: You should build queries in the CI model, not in the views. This violates the basic MVC pattern

File: application/models/reports/Detailed_sales.php [hunk 4]
```diff
@@ -206,5 +213,9 @@ public function getSummaryData(array $inputs)
 
 		return $this->db->get()->row_array();
 	}
+	}  else {
+     echo "";
```
Comment (jekkos) [inline application/models/reports/Detailed_sales.php]: why echo an empty string here?

File: application/views/sales/form.php [hunk 5]
```diff
@@ -43,19 +43,75 @@
...
+			 function SubmitChange(){ var cash = document.getElementById('payment_amount_0'); var debt = document.getElementById('payment_amount_1'); var minus = document.getElementById('VallMinus'); var plus = document.getElementById('VallPlus'); var variable = document.getElementById('minusval'); debt.value = (debt.value - variable.value); cash.value = plus.innerHTML;}</script>";
+			 
+		 if($row1['payment_type'] == 'Cash' || $row1['payment_type'] == 'Debit Card' || $row1['payment_type'] == 'Credit Card' || $row1['payment_type'] == 'Gift Card' || $row1['payment_type'] == 'Check'){
+			 
+?>						
 						<?php echo form_input(array('name'=>'payment_amount_'.$i, 'value'=>$row->payment_amount, 'id'=>'payment_amount_'.$i, 'class'=>'form-control input-sm', 'readonly'=>'true'));?>
+						<input type="text" name="minusval" value="" id="minusval" class="form-control input-sm" oninput="change()">
+						<button onclick="SubmitChange()" class="glyphicon glyphicon-refresh" style="padding: 4px;"></button>
+						<span id="VallMinus" style="display: none;"></span>
```
Comment (jekkos) [inline application/views/sales/form.php]: don't use camelcase in views.. we stick to snake_case for views att his moment

**Comment by mckaygerhard** [comment]
so then 3.2.0 will not have a credit mode payments?

**Comment by mckaygerhard** [comment]
hi @WebShells , i made a work related to this in my cxcteipado branch here: https://github.com/codeigniterpower/osposweb/tree/osposweb-cxcteipado maybe the printing code are not right (i have to move to "receipt" method in sales)

it calculates payment due, made receipts for receive partial payments, also have modified reports and details popup, hard to understand but i hope may help some one that work in this feature.. i hope may help someone.

so then please take a look also @jekkos and @daN4cat i'm not expert but really may help to start a better work.. the comimt history are made "blank" so in that branch only will have the current comits of the work

in lang i only added one line new, so try to reuse all possible but found many bad translations specially in european mades...

**Comment by mckaygerhard** [comment]
better work, related but not similar/specific work: see #2080 (will be realted to xpenses)

**Pull request closed without merging**